draft-kazuho-httpbis-priority.md

```diff
@@ -112,33 +112,93 @@ Unknown parameters MUST be ignored.
-`blocking` first, followed by `document`, followed by `non-blocking`.
+A server SHOULD transmit HTTP responses in the order of their urgency values.
+The lesser the value is, the higher the precedence is.
 
 The following example shows a request for a CSS file with the urgency set to
-`blocking`:
+`-1`:
 
 ~~~ example
 :method = GET
 :scheme = https
 :authority = example.net
 :path = /style.css
-priority = urgency=blocking
+priority = urgency=-1
 ~~~
 
+The definition of the urgencies and their expected use-case are described below.
+Endpoints SHOULD respect the definition of the values when assigning urgencies.
+
+### blocking
+
+The blocking urgency indicates that the response prevents other responses from
+being used.
+
+Value of -1 is used to represent this urgency.
+
+For example, a stylesheet being referred from the HEAD section of an HTML
+document blocks a web browser from rendering the HTML.  In such case, the
+stylesheet is given the blocking urgency.
+
+### document
+
+The document urgency indicates that the response contains the document that is
+being processed.  This urgency is also used for responses that deserve the same
+precedence as the contents of the document.
+
+Value of zero (0) is used to represent this urgency.
+
+For example, when a user using a web browser navigates to a different HTML
+document, the request for that HTML is given the document urgency.  When that
+HTML document uses a custom font, the request for that custom font SHOULD also
+be given the document urgency.  This is because the availablity of the custom
+font is the precondition for the user to use the portion of the HTML document,
+which is to be rendered by that font.
+
+### non-blocking
+
+The non-blocking urgency indicates that the response does not prevent the client
+from using the document even though the response is being incorporated to or
+referred to by the document.
+
+For example, inline images typically improve the usability of an HTML document.
+Therefore, such images will be given the non-blocking urgency.
+
+Values between 1 and 5 are used to represent this urgency, to provide wiggle
+room for endpoints to give some responses more or less precedence than others
+that belong to the non-blocking group.
+
+Clients SHOULD NOT use values 1 and 5.  Servers MAY use these values to
+prioritize a response above or below other non-blocking responses.
+
+Clients MAY use values 2 and 4 to indicate that a request is given relatively
+higher or lower priority within the non-blocking urgency group.
+
+When none of the considerations above is applicable, the value of 3 SHOULD be
+used.
+
+### background
+
+The background urgency is used for responses of which the delivery can be
+postponed without having an impact on using other responses.
+
+Value of 6 is used to represent this urgency.
+
+As an example, a software update running in the background would be assigned the
+background urgency.
```

Comment:
I don't really like this example for the web browsing use case, which you reference later in the document (see the "ordinary web browsing" part). 

Alternative:
As an example, clients can prefetch resources that are expected to be needed for the next page load. As these will not affect the the current page, they would all be assigned the background urgency.

Better with the final addition:
. If these prefetched resources are still in flight when the navigation to the next page happens, they can be re-prioritized to their actual urgency values.

Reply:
Thank you for the suggestion.

On one hand I agree, though on the other hand I would be hesitant to use prefetch as an example without talking about reprioritization in general.

Maybe I'd argue that "ordinary web browsing" include all the tasks that are run in the background, while the user uses the web browser for "ordinary web browsing."

Thoughts?

Reply:
I do indeed feel a bit strongly about this, as I don't think a "software update" makes sense in terms of prioritization on one connection... 

As I see it, software updates would typically happen on a separate connection, not on the same QUIC connection used to load the webpage. As such, you would be talking about cross-connection prioritization/congestion control, which imo is a higher-level issue that doesn't really seem to fit in this text. Unless you envision me downloading an update for Google Chrome while also using google calendar and both endpoints are coalesced onto the same connection or something, but that seems a bit of a stretch and not really the -main- use case here.

Looking at the mail thread where @ianswett and @patmeenan mention the background use cases, it's a bit unclear to me how they see this concretely. They also mention background **uploads**, which also seems a bit weird. Overall, it seems like they envision a browser-wide (or even OS-wide) prioritization setup, managing the full outgoing bandwidth of everything combined. In that case, background obviously makes more sense, but again: doesn't fit in this draft? Maybe they can weigh in here? 


On the other hand, I'm struggling to find good uses for "background" that aren't prefetch. Service Worker (cache) updates are one. I was also thinking about "blog post comments" or "social media plugins" but those are probably best at level 4. 

Overall, as I feel we still need a bit of an updated discussion on re-prioritization in the text down the line anyway, we can/should keep the prefetch use case as the example of background.

Reply:
I get your point.

While I do not think referring to software updates necessarily lead to the conclusion that we are talking about cross-connection prioritization (they could be coalesced onto another connection that goes to Google), I think you are correct in pointing out that it would be rare to see software update running on a connection that you'd be using for web browsing. Because software update only connects to a specific server.

Reply:
I had been meaning to make a similar point to @rmarx here. As a group we have discussed some use cases that are interested but the theme of this proposal is grounded in a web browser activity. Not all network activity from a browser directly related to the navigation or user-facing interaction. While some of these activities might get squeezed into the same connection - coalescing certainly might make that happen, perhaps multipath too? - it seems like somethign on the fringe of this spec's primary purpose.